{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "336f79f8-54b1-4ac1-aeca-5557e88516b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "from snakemake.io import expand\n",
    "import pandas as pd\n",
    "import yaml\n",
    "import pdb\n",
    "import re\n",
    "import numpy as np\n",
    "import pyranges as pr\n",
    "from utils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "15ed8589-15fc-448b-951f-6f7cb10fb9da",
   "metadata": {},
   "outputs": [],
   "source": [
    "config_file = 'config.yml'\n",
    "with open(config_file) as f:\n",
    "    config = yaml.safe_load(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c18623d-9dcf-495b-837f-26ea7815d4e2",
   "metadata": {},
   "source": [
    "## 230627 get the chr restructuring right"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "57facaf1-189c-4609-ab88-57ab559118e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "gff_file = 'data/human/raw/Placenta_Placenta_ont_pre-capture_human_all.gff'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "80d14b30-5bff-4d38-ab16-0932ed50b42d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gff_rm_sirv(gff_file, 'test')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d579a31-15c4-4009-9feb-6e1b362aa541",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbb6ae15-137c-403d-9511-3a22fc38572e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e46f971-7c73-4018-aba7-8d16f3513a0e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2872bd5-280f-4eaa-a7c0-a9fc53a939e0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cafd9dd-52ae-49b8-8fed-7a7393f24c3d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "817323e6-8bc5-409f-9d97-c60581ad637f",
   "metadata": {},
   "source": [
    "## 230626 which chromosomes are in the reference fasta and not in the input gffs (or vice vers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "67ea3354-e416-407b-80fc-7dfafb944d5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = pr.read_gff('data/human/Liver_Adult_pacBioSII_pre-capture_human_all.gff').as_df()\n",
    "import glob\n",
    "import pyranges as pr\n",
    "\n",
    "chrs = []\n",
    "for f in glob.glob('data/human/raw/*gff'):\n",
    "    print(f)\n",
    "    df = pr.read_gff(f).as_df()\n",
    "    chrs += df.Chromosome.unique().tolist()\n",
    "    chrs = list(set(chrs))\n",
    "    \n",
    "    \n",
    "with open('chrs.txt', 'w') as ofile:\n",
    "    ofile.write(','.join(chrs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "73dcd590-552c-478c-8732-ed665427fc1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# chrs = df.Chromosome.unique().tolist()\n",
    "# chrs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6cde2711-2ab5-40ec-b676-3a7ffb65e178",
   "metadata": {},
   "outputs": [],
   "source": [
    "chrs = ['KI270742.1','KI270746.1','GL000224.1','KI270719.1','KI270743.1','KI270717.1','GL000213.1',\\\n",
    "        'KI270754.1','chr20','KI270718.1','chr5','GL000009.2','KI270733.1','GL000205.2','KI270745.1',\\\n",
    "        'GL000220.1','KI270750.1','chr3','KI270721.1','KI270713.1','KI270720.1',\\\n",
    "        'GL000218.1','KI270726.1','chr1','SIRVome_isoforms','chr2','chr12','chrM',\\\n",
    "        'KI270734.1','GL000225.1','KI270755.1','GL000008.2','KI270706.1','chr17','KI270727.1',\\\n",
    "        'chr13','chr19','chr21','KI270722.1','KI270749.1','chr11','KI270751.1','chrY','chr16',\\\n",
    "        'KI270711.1','GL000214.1','KI270714.1','chr8','chr4','GL000221.1','chr9','chrX','KI270744.1',\\\n",
    "        'chr15','KI270753.1','KI270731.1','chr6','chr18','KI270725.1','GL000194.1','KI270712.1','chr7',\\\n",
    "        'KI270442.1','GL000195.1','KI270748.1','chr14','KI270741.1','KI270728.1','GL000219.1','chr10','KI270710.1','chr22']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "65d98ceb-fa7d-433c-9bc3-9a905f567015",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KI270742.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KI270746.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GL000224.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KI270719.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KI270743.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome\n",
       "0  KI270742.1\n",
       "1  KI270746.1\n",
       "2  GL000224.1\n",
       "3  KI270719.1\n",
       "4  KI270743.1"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_df = pd.DataFrame(data=chrs, columns=['Chromosome'])\n",
    "gff_df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "5919deb6-a707-4219-8f79-726e58f90961",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ! gzcat ~/Desktop/hg38.fa.gz | grep '^>' > human_ref_chrs.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "89479699-767c-421d-b498-a7e52559a424",
   "metadata": {},
   "outputs": [],
   "source": [
    "fa_df = pd.read_csv('human_ref_chrs.txt',header=None)\n",
    "fa_df.columns = ['chr']\n",
    "fa_df['Chromosome'] = fa_df.chr.str.split('>', n=1, expand=True)[1]\n",
    "fa_df = fa_df[['Chromosome']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "51adf48f-1693-4310-918b-1088d75f1620",
   "metadata": {},
   "outputs": [],
   "source": [
    "fa_df\n",
    "fa_chrs = fa_df.Chromosome.unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "67142da8-729b-4726-a612-b33a652b4b69",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['chrUn_KI270374v1',\n",
       " 'chrUn_KI270373v1',\n",
       " 'chr11_KI270721v1_random',\n",
       " 'chrUn_KI270466v1',\n",
       " 'chrUn_KI270418v1',\n",
       " 'chrUn_KI270387v1',\n",
       " 'chrUn_KI270579v1',\n",
       " 'chr16_KI270856v1_alt',\n",
       " 'chrUn_KI270747v1',\n",
       " 'chr17_GL383564v2_alt',\n",
       " 'chr17_KI270857v1_alt',\n",
       " 'chr12_KI270834v1_alt',\n",
       " 'chr19_KI270922v1_alt',\n",
       " 'chrUn_KI270749v1',\n",
       " 'chr14_KI270724v1_random',\n",
       " 'chr22_KI270731v1_random',\n",
       " 'chr9_KI270823v1_alt',\n",
       " 'chr16_KI270854v1_alt',\n",
       " 'chr7_KI270807v1_alt',\n",
       " 'chr19_GL949749v2_alt',\n",
       " 'chr12_KI270904v1_alt',\n",
       " 'chr6_GL000252v2_alt',\n",
       " 'chr14_GL000194v1_random',\n",
       " 'chr18_GL383570v1_alt',\n",
       " 'chrUn_GL000224v1',\n",
       " 'chrUn_KI270322v1',\n",
       " 'chrUn_KI270529v1',\n",
       " 'chr19_GL949748v2_alt',\n",
       " 'chr1_KI270761v1_alt',\n",
       " 'chr9_KI270717v1_random',\n",
       " 'chr5_KI270793v1_alt',\n",
       " 'chr22_KI270928v1_alt',\n",
       " 'chr3_KI270895v1_alt',\n",
       " 'chrUn_KI270743v1',\n",
       " 'chr19_KI270891v1_alt',\n",
       " 'chr16_KI270855v1_alt',\n",
       " 'chr4_KI270789v1_alt',\n",
       " 'chrUn_GL000214v1',\n",
       " 'chrUn_GL000219v1',\n",
       " 'chrUn_KI270507v1',\n",
       " 'chrUn_KI270750v1',\n",
       " 'chrUn_KI270756v1',\n",
       " 'chr2_KI270773v1_alt',\n",
       " 'chr9_KI270718v1_random',\n",
       " 'chr8_KI270811v1_alt',\n",
       " 'chr18_GL383572v1_alt',\n",
       " 'chrUn_KI270519v1',\n",
       " 'chrUn_KI270392v1',\n",
       " 'chr3_JH636055v2_alt',\n",
       " 'chr5_KI270791v1_alt',\n",
       " 'chr19_KI270916v1_alt',\n",
       " 'chr15_KI270849v1_alt',\n",
       " 'chr22_KB663609v1_alt',\n",
       " 'chr12_GL383549v1_alt',\n",
       " 'chr8_KI270817v1_alt',\n",
       " 'chr11_KI270832v1_alt',\n",
       " 'chrUn_KI270391v1',\n",
       " 'chr9_GL383542v1_alt',\n",
       " 'chr22_KI270877v1_alt',\n",
       " 'chr11_KI270829v1_alt',\n",
       " 'chr14_KI270847v1_alt',\n",
       " 'chr19_KI270931v1_alt',\n",
       " 'chrUn_KI270742v1',\n",
       " 'chr6_GL000251v2_alt',\n",
       " 'chr8_KI270816v1_alt',\n",
       " 'chr12_GL383552v1_alt',\n",
       " 'chr20_KI270869v1_alt',\n",
       " 'chrUn_KI270379v1',\n",
       " 'chr17_KI270909v1_alt',\n",
       " 'chrUn_KI270382v1',\n",
       " 'chr1_KI270711v1_random',\n",
       " 'chr5_KI270898v1_alt',\n",
       " 'chr14_KI270725v1_random',\n",
       " 'chr14_GL000225v1_random',\n",
       " 'chrUn_KI270336v1',\n",
       " 'chrUn_KI270511v1',\n",
       " 'chrUn_KI270364v1',\n",
       " 'chrUn_KI270754v1',\n",
       " 'chr21_KI270872v1_alt',\n",
       " 'chrUn_KI270302v1',\n",
       " 'chrUn_KI270394v1',\n",
       " 'chr1_KI270764v1_alt',\n",
       " 'chr19_GL949751v2_alt',\n",
       " 'chr22_KI270735v1_random',\n",
       " 'chr6_GL000253v2_alt',\n",
       " 'chr14_KI270844v1_alt',\n",
       " 'chrUn_KI270512v1',\n",
       " 'chr18_KI270911v1_alt',\n",
       " 'chrUn_KI270381v1',\n",
       " 'chrUn_KI270548v1',\n",
       " 'chr17_KI270859v1_alt',\n",
       " 'chr1_GL383520v2_alt',\n",
       " 'chr16_GL383556v1_alt',\n",
       " 'chr17_GL383566v1_alt',\n",
       " 'chrUn_KI270383v1',\n",
       " 'chrUn_KI270412v1',\n",
       " 'chr12_GL383553v2_alt',\n",
       " 'chr19_KI270882v1_alt',\n",
       " 'chr7_KI270804v1_alt',\n",
       " 'chr2_KI270769v1_alt',\n",
       " 'chr19_GL383574v1_alt',\n",
       " 'chr7_KI270805v1_alt',\n",
       " 'chr5_GL000208v1_random',\n",
       " 'chr4_GL000257v2_alt',\n",
       " 'chr4_KI270896v1_alt',\n",
       " 'chr12_KI270833v1_alt',\n",
       " 'chrUn_KI270419v1',\n",
       " 'chr5_GL383531v1_alt',\n",
       " 'chr2_KI270716v1_random',\n",
       " 'chr18_KI270912v1_alt',\n",
       " 'chrUn_KI270389v1',\n",
       " 'chr3_KI270934v1_alt',\n",
       " 'chr8_KI270901v1_alt',\n",
       " 'chr1_KI270706v1_random',\n",
       " 'chrUn_KI270593v1',\n",
       " 'chrUn_KI270316v1',\n",
       " 'chr9_KI270719v1_random',\n",
       " 'chrUn_KI270522v1',\n",
       " 'chr1_KI270710v1_random',\n",
       " 'chr22_GL383582v2_alt',\n",
       " 'chr21_GL383580v2_alt',\n",
       " 'chr22_GL383583v2_alt',\n",
       " 'chr3_KI270784v1_alt',\n",
       " 'chr11_KI270903v1_alt',\n",
       " 'chr7_GL383534v2_alt',\n",
       " 'chr16_KI270728v1_random',\n",
       " 'chr7_KI270803v1_alt',\n",
       " 'chrUn_KI270530v1',\n",
       " 'chr2_KI270776v1_alt',\n",
       " 'chr19_KI270921v1_alt',\n",
       " 'chrUn_KI270745v1',\n",
       " 'chr1_KI270892v1_alt',\n",
       " 'chr12_KI270836v1_alt',\n",
       " 'chrUn_KI270746v1',\n",
       " 'chr18_KI270863v1_alt',\n",
       " 'chr19_GL949752v1_alt',\n",
       " 'chrUn_KI270422v1',\n",
       " 'chr19_GL383576v1_alt',\n",
       " 'chrUn_GL000195v1',\n",
       " 'chr21_GL383581v2_alt',\n",
       " 'chr19_GL383573v1_alt',\n",
       " 'chr16_KI270853v1_alt',\n",
       " 'chr3_KI270783v1_alt',\n",
       " 'chrUn_KI270753v1',\n",
       " 'chr19_KI270923v1_alt',\n",
       " 'chrUn_KI270757v1',\n",
       " 'chr1_KI270760v1_alt',\n",
       " 'chr3_KI270924v1_alt',\n",
       " 'chrUn_KI270539v1',\n",
       " 'chrUn_KI270386v1',\n",
       " 'chr19_GL383575v2_alt',\n",
       " 'chr22_KI270876v1_alt',\n",
       " 'chr19_KI270887v1_alt',\n",
       " 'chrUn_KI270429v1',\n",
       " 'chr6_KI270798v1_alt',\n",
       " 'chr19_KI270919v1_alt',\n",
       " 'chrUn_KI270330v1',\n",
       " 'chr8_KI270926v1_alt',\n",
       " 'chr2_KI270715v1_random',\n",
       " 'chrUn_KI270744v1',\n",
       " 'chr15_KI270727v1_random',\n",
       " 'chr17_GL000205v2_random',\n",
       " 'chrUn_KI270591v1',\n",
       " 'chr7_KI270809v1_alt',\n",
       " 'chr6_GL000256v2_alt',\n",
       " 'chr22_KI270734v1_random',\n",
       " 'chr1_KI270765v1_alt',\n",
       " 'chr2_KI270893v1_alt',\n",
       " 'chr7_KI270808v1_alt',\n",
       " 'chrUn_KI270583v1',\n",
       " 'chr18_GL383567v1_alt',\n",
       " 'chrUn_KI270588v1',\n",
       " 'chrY_KI270740v1_random',\n",
       " 'chr19_KI270889v1_alt',\n",
       " 'chr5_KI270792v1_alt',\n",
       " 'chr12_KI270837v1_alt',\n",
       " 'chrUn_KI270752v1',\n",
       " 'chr1_KI270712v1_random',\n",
       " 'chr21_KI270873v1_alt',\n",
       " 'chr3_KI270778v1_alt',\n",
       " 'chrUn_KI270590v1',\n",
       " 'chr13_KI270838v1_alt',\n",
       " 'chr15_KI270850v1_alt',\n",
       " 'chr14_KI270723v1_random',\n",
       " 'chr1_KI270762v1_alt',\n",
       " 'chr17_KI270862v1_alt',\n",
       " 'chrUn_KI270311v1',\n",
       " 'chr9_KI270720v1_random',\n",
       " 'chr6_KI270797v1_alt',\n",
       " 'chrUn_KI270435v1',\n",
       " 'chr5_GL383530v1_alt',\n",
       " 'chrUn_KI270385v1',\n",
       " 'chr1_KI270713v1_random',\n",
       " 'chr19_KI270914v1_alt',\n",
       " 'chr2_KI270775v1_alt',\n",
       " 'chr21_KI270874v1_alt',\n",
       " 'chrUn_KI270538v1',\n",
       " 'chr8_KI270819v1_alt',\n",
       " 'chr18_GL383568v1_alt',\n",
       " 'chr8_KI270818v1_alt',\n",
       " 'chr1_GL383519v1_alt',\n",
       " 'chr19_KI270918v1_alt',\n",
       " 'chr19_KI270930v1_alt',\n",
       " 'chrUn_KI270371v1',\n",
       " 'chrUn_KI270411v1',\n",
       " 'chr10_GL383545v1_alt',\n",
       " 'chr2_KI270894v1_alt',\n",
       " 'chrUn_KI270366v1',\n",
       " 'chr3_KI270935v1_alt',\n",
       " 'chrX_KI270880v1_alt',\n",
       " 'chr5_KI270794v1_alt',\n",
       " 'chr5_KI270897v1_alt',\n",
       " 'chr11_KI270902v1_alt',\n",
       " 'chr1_KI270707v1_random',\n",
       " 'chr1_KI270766v1_alt',\n",
       " 'chrUn_KI270304v1',\n",
       " 'chrUn_KI270515v1',\n",
       " 'chr5_GL339449v2_alt',\n",
       " 'chr8_KI270810v1_alt',\n",
       " 'chr12_GL877876v1_alt',\n",
       " 'chr4_KI270788v1_alt',\n",
       " 'chr22_KI270879v1_alt',\n",
       " 'chrUn_KI270517v1',\n",
       " 'chr2_KI270767v1_alt',\n",
       " 'chr21_GL383578v2_alt',\n",
       " 'chr12_GL877875v1_alt',\n",
       " 'chr18_GL383569v1_alt',\n",
       " 'chr4_KI270790v1_alt',\n",
       " 'chr19_KI270885v1_alt',\n",
       " 'chr5_KI270796v1_alt',\n",
       " 'chrUn_KI270303v1',\n",
       " 'chrUn_GL000213v1',\n",
       " 'chrUn_KI270420v1',\n",
       " 'chr17_KI270729v1_random',\n",
       " 'chr8_KI270813v1_alt',\n",
       " 'chr11_KI270927v1_alt',\n",
       " 'chr8_KI270812v1_alt',\n",
       " 'chr11_JH159136v1_alt',\n",
       " 'chr5_GL383532v1_alt',\n",
       " 'chrUn_GL000226v1',\n",
       " 'chrUn_KI270587v1',\n",
       " 'chr8_KI270815v1_alt',\n",
       " 'chr6_GL000254v2_alt',\n",
       " 'chrUn_KI270425v1',\n",
       " 'chr19_KI270883v1_alt',\n",
       " 'chr22_KI270878v1_alt',\n",
       " 'chr4_GL000008v2_random',\n",
       " 'chrUn_KI270317v1',\n",
       " 'chr3_KI270782v1_alt',\n",
       " 'chr15_KI270852v1_alt',\n",
       " 'chr8_KI270822v1_alt',\n",
       " 'chr6_GL000255v2_alt',\n",
       " 'chr11_KI270830v1_alt',\n",
       " 'chrUn_KI270376v1',\n",
       " 'chr15_GL383555v2_alt',\n",
       " 'chrUn_KI270748v1',\n",
       " 'chr2_KI270771v1_alt',\n",
       " 'chr4_GL383527v1_alt',\n",
       " 'chrUn_KI270751v1',\n",
       " 'chr14_KI270722v1_random',\n",
       " 'chrUn_KI270468v1',\n",
       " 'chr17_KI270730v1_random',\n",
       " 'chr20_KI270871v1_alt',\n",
       " 'chrUn_KI270584v1',\n",
       " 'chr12_KI270835v1_alt',\n",
       " 'chr14_KI270846v1_alt',\n",
       " 'chr19_KI270920v1_alt',\n",
       " 'chrUn_KI270438v1',\n",
       " 'chr19_KI270866v1_alt',\n",
       " 'chr11_KI270831v1_alt',\n",
       " 'chrUn_KI270589v1',\n",
       " 'chrUn_KI270335v1',\n",
       " 'chr14_KI270726v1_random',\n",
       " 'chr3_KI270779v1_alt',\n",
       " 'chr11_JH159137v1_alt',\n",
       " 'chr17_KI270910v1_alt',\n",
       " 'chrX_KI270881v1_alt',\n",
       " 'chr1_GL383518v1_alt',\n",
       " 'chr6_GL000250v2_alt',\n",
       " 'chrUn_KI270334v1',\n",
       " 'chr19_KI270868v1_alt',\n",
       " 'chrUn_GL000216v2',\n",
       " 'chr19_KI270888v1_alt',\n",
       " 'chr19_KI270917v1_alt',\n",
       " 'chr6_KI270800v1_alt',\n",
       " 'chr16_GL383557v1_alt',\n",
       " 'chr19_GL949750v2_alt',\n",
       " 'chrUn_KI270372v1',\n",
       " 'chr6_GL383533v1_alt',\n",
       " 'chrUn_KI270396v1',\n",
       " 'chrUn_KI270338v1',\n",
       " 'chr19_KI270915v1_alt',\n",
       " 'chr19_GL949753v2_alt',\n",
       " 'chr17_JH159148v1_alt',\n",
       " 'chr3_KI270937v1_alt',\n",
       " 'chrUn_GL000220v1',\n",
       " 'chrUn_GL000218v1',\n",
       " 'chrUn_KI270363v1',\n",
       " 'chrUn_KI270320v1',\n",
       " 'chrUn_KI270337v1',\n",
       " 'chrUn_KI270509v1',\n",
       " 'chr2_KI270774v1_alt',\n",
       " 'chr9_GL383540v1_alt',\n",
       " 'chr17_JH159146v1_alt',\n",
       " 'chr2_KI270768v1_alt',\n",
       " 'chrUn_KI270442v1',\n",
       " 'chr2_KI270770v1_alt',\n",
       " 'chr6_KB021644v2_alt',\n",
       " 'chr17_KI270860v1_alt',\n",
       " 'chrUn_KI270393v1',\n",
       " 'chr8_KI270821v1_alt',\n",
       " 'chr22_KI270736v1_random',\n",
       " 'chr19_KI270929v1_alt',\n",
       " 'chrUn_KI270508v1',\n",
       " 'chr10_KI270824v1_alt',\n",
       " 'chr3_KI270936v1_alt',\n",
       " 'chr19_KI270938v1_alt',\n",
       " 'chr11_KI270827v1_alt',\n",
       " 'chrUn_KI270395v1',\n",
       " 'chrUn_KI270423v1',\n",
       " 'chrUn_KI270580v1',\n",
       " 'chr21_GL383579v2_alt',\n",
       " 'chr9_GL383539v1_alt',\n",
       " 'chr7_KI270806v1_alt',\n",
       " 'chr13_KI270843v1_alt',\n",
       " 'chr15_KI270851v1_alt',\n",
       " 'chr4_KI270786v1_alt',\n",
       " 'chr4_GL383528v1_alt',\n",
       " 'chr15_KI270848v1_alt',\n",
       " 'chr22_KI270737v1_random',\n",
       " 'chr19_GL000209v2_alt',\n",
       " 'chr6_KI270758v1_alt',\n",
       " 'chr1_KI270709v1_random',\n",
       " 'chrUn_KI270510v1',\n",
       " 'chrUn_KI270417v1',\n",
       " 'chr15_KI270905v1_alt',\n",
       " 'chrUn_KI270315v1',\n",
       " 'chr15_GL383554v1_alt',\n",
       " 'chr17_KI270861v1_alt',\n",
       " 'chrUn_KI270544v1',\n",
       " 'chr17_GL383563v3_alt',\n",
       " 'chr6_KI270802v1_alt',\n",
       " 'chr14_GL000009v2_random',\n",
       " 'chr3_KI270781v1_alt',\n",
       " 'chr22_KI270732v1_random',\n",
       " 'chr4_KI270925v1_alt',\n",
       " 'chr13_KI270840v1_alt',\n",
       " 'chr19_KI270890v1_alt',\n",
       " 'chr11_GL383547v1_alt',\n",
       " 'chrUn_KI270384v1',\n",
       " 'chr17_JH159147v1_alt',\n",
       " 'chrUn_KI270390v1',\n",
       " 'chr22_KI270875v1_alt',\n",
       " 'chrUn_KI270375v1',\n",
       " 'chr19_KI270867v1_alt',\n",
       " 'chrUn_KI270448v1',\n",
       " 'chrUn_KI270518v1',\n",
       " 'chr17_GL000258v2_alt',\n",
       " 'chr13_KI270841v1_alt',\n",
       " 'chr8_KI270900v1_alt',\n",
       " 'chrUn_KI270582v1',\n",
       " 'chrUn_KI270312v1',\n",
       " 'chrUn_KI270310v1',\n",
       " 'chrUn_KI270388v1',\n",
       " 'chr13_KI270842v1_alt',\n",
       " 'chr1_KI270759v1_alt',\n",
       " 'chr1_KI270763v1_alt',\n",
       " 'chr17_GL383565v1_alt',\n",
       " 'chr19_KI270886v1_alt',\n",
       " 'chr22_KI270739v1_random',\n",
       " 'chr17_KI270908v1_alt',\n",
       " 'chr3_KI270777v1_alt',\n",
       " 'chr20_KI270870v1_alt',\n",
       " 'chrUn_KI270424v1',\n",
       " 'chr14_KI270845v1_alt',\n",
       " 'chr17_KI270858v1_alt',\n",
       " 'chrUn_KI270741v1',\n",
       " 'chr2_GL383522v1_alt',\n",
       " 'chr12_GL383550v2_alt',\n",
       " 'chrUn_KI270581v1',\n",
       " 'chr6_KI270801v1_alt',\n",
       " 'chrX_KI270913v1_alt',\n",
       " 'chrUn_KI270378v1',\n",
       " 'chr19_KI270933v1_alt',\n",
       " 'chrUn_KI270467v1',\n",
       " 'chrUn_KI270305v1',\n",
       " 'chr3_KI270780v1_alt',\n",
       " 'chr3_GL383526v1_alt',\n",
       " 'chrUn_KI270516v1',\n",
       " 'chrUn_KI270333v1',\n",
       " 'chr18_GL383571v1_alt',\n",
       " 'chr19_KI270865v1_alt',\n",
       " 'chr7_KI270899v1_alt',\n",
       " 'chr19_KI270932v1_alt',\n",
       " 'chrUn_KI270329v1',\n",
       " 'chr22_KI270733v1_random',\n",
       " 'chr5_KI270795v1_alt',\n",
       " 'chr17_KI270907v1_alt',\n",
       " 'chr10_GL383546v1_alt',\n",
       " 'chrUn_KI270755v1',\n",
       " 'chr4_KI270787v1_alt',\n",
       " 'chrUn_KI270521v1',\n",
       " 'chrUn_KI270528v1',\n",
       " 'chr18_KI270864v1_alt',\n",
       " 'chr15_KI270906v1_alt',\n",
       " 'chr19_GL949746v1_alt',\n",
       " 'chr3_GL000221v1_random',\n",
       " 'chr2_KI270772v1_alt',\n",
       " 'chr19_KI270884v1_alt',\n",
       " 'chr1_KI270708v1_random',\n",
       " 'chr8_KI270820v1_alt',\n",
       " 'chr20_GL383577v2_alt',\n",
       " 'chr8_KI270814v1_alt',\n",
       " 'chr22_KI270738v1_random',\n",
       " 'chr9_GL383541v1_alt',\n",
       " 'chr12_GL383551v1_alt',\n",
       " 'chrUn_KI270465v1',\n",
       " 'chr5_GL949742v1_alt',\n",
       " 'chr4_KI270785v1_alt',\n",
       " 'chr19_GL949747v2_alt',\n",
       " 'chr11_KI270826v1_alt',\n",
       " 'chrUn_KI270362v1',\n",
       " 'chrUn_KI270414v1',\n",
       " 'chr2_GL582966v2_alt',\n",
       " 'chr13_KI270839v1_alt',\n",
       " 'chr10_KI270825v1_alt',\n",
       " 'chr2_GL383521v1_alt',\n",
       " 'chrUn_KI270340v1',\n",
       " 'chr6_KI270799v1_alt',\n",
       " 'chr1_KI270714v1_random']"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(set(fa_chrs)-set(chrs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f34c8861-f8b7-45fc-8ff6-23255481e682",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['KI270711.1',\n",
       " 'KI270726.1',\n",
       " 'KI270748.1',\n",
       " 'KI270749.1',\n",
       " 'GL000213.1',\n",
       " 'GL000221.1',\n",
       " 'KI270742.1',\n",
       " 'GL000218.1',\n",
       " 'KI270745.1',\n",
       " 'KI270744.1',\n",
       " 'GL000205.2',\n",
       " 'KI270743.1',\n",
       " 'KI270722.1',\n",
       " 'KI270706.1',\n",
       " 'KI270731.1',\n",
       " 'KI270733.1',\n",
       " 'KI270755.1',\n",
       " 'KI270720.1',\n",
       " 'GL000224.1',\n",
       " 'KI270746.1',\n",
       " 'KI270719.1',\n",
       " 'KI270727.1',\n",
       " 'KI270753.1',\n",
       " 'KI270710.1',\n",
       " 'SIRVome_isoforms',\n",
       " 'GL000220.1',\n",
       " 'KI270721.1',\n",
       " 'KI270741.1',\n",
       " 'KI270751.1',\n",
       " 'GL000225.1',\n",
       " 'KI270750.1',\n",
       " 'KI270725.1',\n",
       " 'KI270442.1',\n",
       " 'KI270712.1',\n",
       " 'KI270734.1',\n",
       " 'GL000008.2',\n",
       " 'KI270713.1',\n",
       " 'GL000194.1',\n",
       " 'KI270718.1',\n",
       " 'KI270717.1',\n",
       " 'KI270754.1',\n",
       " 'GL000219.1',\n",
       " 'KI270714.1',\n",
       " 'GL000195.1',\n",
       " 'KI270728.1',\n",
       " 'GL000214.1',\n",
       " 'GL000009.2']"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(set(chrs)-set(fa_chrs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "8711caa2-0aba-4bd2-9322-3d06dfd5854d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "chr_map = defaultdict(list)\n",
    "for c in chrs:\n",
    "    if c not in fa_chrs:\n",
    "        for fa_c in fa_chrs:\n",
    "            thing = c.split('.')[0]\n",
    "            if thing in fa_c:\n",
    "                chr_map[c].append(fa_c)\n",
    "    else:\n",
    "        chr_map[c].append(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "2e842927-6012-4666-bd0f-793dee116cc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# chr_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "ec3b20fe-3168-49ae-8ae3-c500c4434fa9",
   "metadata": {},
   "outputs": [],
   "source": [
    "for key, item in chr_map.items():\n",
    "    assert len(item) == 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "62fbdc7c-72f5-4c89-b5c3-b075d5aac373",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame.from_dict(chr_map, orient='index')\n",
    "df.reset_index(inplace=True)\n",
    "df.rename({'index':'gff_chr', 0:'fa_chr'}, axis=1, inplace=True)\n",
    "\n",
    "df.to_csv('chr_map.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31751e45-9e31-4c2e-b6c3-746e61ffdd8e",
   "metadata": {},
   "source": [
    "## 230623 seq. cerb annot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "0f45f88b-8d48-4c1d-9322-7277fa7bc822",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "88\n"
     ]
    }
   ],
   "source": [
    "config_fname = '230614_config.tsv'\n",
    "df = parse_input_config(config, config_fname, 'all')\n",
    "\n",
    "# todo\n",
    "df = df.loc[df.species == 'mouse']\n",
    "\n",
    "datasets = df.dataset.tolist()\n",
    "species = df.species.tolist()\n",
    "max_cerberus_run = len(datasets) # TODO modify to work w/ multiple species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "57725824-8b37-4f01-b528-c0ac92969ecc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "40"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max_cerberus_run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eddd3be2-29eb-4bcc-b4d6-631faaa41532",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "108cebe9-6555-4e67-940f-ddf84c528ccf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3da4b066-bbd7-4c15-854a-cb5b1caf0966",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "522631dc-0adf-4324-82c8-1db21b7560f4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b4dbc2c-0f5b-49bb-a4f9-b82a4c5a5606",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "350af123-0e83-4bce-8e6e-c5544560a4a3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "bf1bd7be-c4b7-473c-ab99-2715110436bb",
   "metadata": {},
   "source": [
    "## 230621 why no ref ends"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "00eadcf6-c455-4f3a-991c-2fed430dc8ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "88\n"
     ]
    }
   ],
   "source": [
    "# data config\n",
    "from utils import *\n",
    "\n",
    "config_fname = '230614_config.tsv'\n",
    "df = parse_input_config(config, config_fname, 'all')\n",
    "df = df.loc[df.species == 'mouse']\n",
    "\n",
    "datasets = df.dataset.tolist()\n",
    "\n",
    "def get_df_col(df, q_col, q_col_val, t_col):\n",
    "    return df.loc[df[q_col]==q_col_val, t_col].tolist()\n",
    "\n",
    "def get_agg_settings(end_mode, species, param='files'):\n",
    "    settings = {}\n",
    "    files = []\n",
    "    sources = []\n",
    "\n",
    "    if 'end_mode':\n",
    "        ic = False\n",
    "    else:\n",
    "        ic = True\n",
    "\n",
    "    # get reference ics / tsss / tess first\n",
    "    # first source is 'cerberus' to indicate\n",
    "    # preservation of source / novelty\n",
    "    if ic == True:\n",
    "        files += [config['ref']['ics']]\n",
    "    else:\n",
    "        files += expand(config['ref']['ends'],\n",
    "                        zip,\n",
    "                        end_mode=end_mode,\n",
    "                        species=species)\n",
    "    sources += ['gencode']\n",
    "\n",
    "    # then get the files from each study\n",
    "    # use the datasets as the sources\n",
    "    if ic == True:\n",
    "        files += expand(expand(config['data']['cerb']['ics'],\n",
    "               zip,\n",
    "               dataset=get_df_col(df, 'species', species, 'dataset'),\n",
    "               allow_missing=True),\n",
    "               species=species)\n",
    "    else:\n",
    "        files += expand(expand(config['data']['cerb']['ends'],\n",
    "               zip,\n",
    "               dataset=get_df_col(df, 'species', species, 'dataset'),\n",
    "               allow_missing=True),\n",
    "               species=species,\n",
    "               end_mode=end_mode)\n",
    "    sources += datasets\n",
    "\n",
    "    settings['file'] = files\n",
    "    settings['source'] = sources\n",
    "\n",
    "    return settings[param]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "58a8faa2-04a9-4376-a238-84285a8d36d8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'data/mouse/cerberus/agg_tss.bed'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expand(config['data']['cerb']['agg_ends'],\n",
    "           species='mouse',\n",
    "           end_mode='tss')[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "47df892f-f95b-4a91-b413-cfdb456f6f5d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'data/mouse/cerberus/agg_tes.bed'"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expand(config['data']['cerb']['agg_ends'],\n",
    "                              species='mouse',\n",
    "                              end_mode='tes')[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6d249d74-a5b5-4dcf-a2ce-789db18b77d8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ref/mouse/tss.bed',\n",
       " 'data/mouse/cerberus/ESC_Embryo_pacBioSII_pre-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/ESC_Embryo_pacBioSII_post-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/ESC_Embryo_ont_pre-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/ESC_Embryo_ont_post-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/WBlood_Adult_pacBioSII_pre-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/WBlood_Adult_pacBioSII_post-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/WBlood_Adult_ont_pre-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/WBlood_Adult_ont_post-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/TpoolA_Adult_pacBioSII_pre-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/TpoolA_Adult_pacBioSII_post-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/TpoolA_Adult_ont_pre-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/TpoolA_Adult_ont_post-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/Testis_Adult_pacBioSII_pre-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/Testis_Adult_pacBioSII_post-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/Testis_Adult_ont_pre-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/Testis_Adult_ont_post-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/Liver_Embryo_pacBioSII_pre-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/Liver_Embryo_pacBioSII_post-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/Liver_Embryo_ont_pre-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/Liver_Embryo_ont_post-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/Liver_Adult_pacBioSII_pre-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/Liver_Adult_pacBioSII_post-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/Liver_Adult_ont_pre-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/Liver_Adult_ont_post-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/Heart_Embryo_pacBioSII_pre-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/Heart_Embryo_pacBioSII_post-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/Heart_Embryo_ont_pre-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/Heart_Embryo_ont_post-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/Heart_Adult_pacBioSII_pre-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/Heart_Adult_pacBioSII_post-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/Heart_Adult_ont_pre-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/Heart_Adult_ont_post-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/Brain_Embryo_pacBioSII_pre-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/Brain_Embryo_pacBioSII_post-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/Brain_Embryo_ont_pre-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/Brain_Embryo_ont_post-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/Brain_Adult_pacBioSII_pre-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/Brain_Adult_pacBioSII_post-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/Brain_Adult_ont_pre-capture_mouse_all/tss.bed',\n",
       " 'data/mouse/cerberus/Brain_Adult_ont_post-capture_mouse_all/tss.bed']"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_agg_settings('tss', 'mouse', 'file')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72472c3e-2b4a-4ca9-8583-4e5ff26e45db",
   "metadata": {},
   "source": [
    "## 230620 WHHHHYYY SNAKEMAKE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d8fd1ce4-56d6-432b-bed3-888ba8c27367",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11\n"
     ]
    }
   ],
   "source": [
    "bad_env = ['/data/homezvol1/freese/miniconda3/envs/SQANTI3.env',\\\n",
    "           '/data/homezvol1/freese/miniconda3/envs/SQANTI3.env/lib/python310.zip',\\\n",
    "           '/data/homezvol1/freese/miniconda3/envs/SQANTI3.env/lib/python3.10',\\\n",
    "           '/data/homezvol1/freese/miniconda3/envs/SQANTI3.env/lib/python3.10/lib-dynload',\\\n",
    "           '/data/homezvol1/freese/.local/lib/python3.10/site-packages',\\\n",
    "           '/dfs8/pub/freese/mortazavi_lab/bin/gencode_utr_fix',\\\n",
    "           '/data/homezvol1/freese/miniconda3/envs/SQANTI3.env/lib/python3.10/site-packages',\\\n",
    "           '/data/homezvol1/freese/miniconda3/envs/SQANTI3.env/lib/python3.10/site-packages/bcbio_gff-0.6.6-py2.7.egg',\\\n",
    "           '/data/homezvol1/freese/miniconda3/envs/SQANTI3.env/lib/python3.10/site-packages/cupcake-29.0.0-py3.10-linux-x86_64.egg',\\\n",
    "           '/data/homezvol1/freese/miniconda3/envs/SQANTI3.env/lib/python3.10/site-packages/sklearn-0.0.post5-py3.10.egg',\\\n",
    "           '/data/homezvol1/freese/miniconda3/envs/SQANTI3.env/lib/python3.10/site-packages/bcbio_gff-0.7.0-py3.10.egg']\n",
    "print(len(bad_env))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e693285d-865d-4a09-baa9-1bc703aacfc0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14\n"
     ]
    }
   ],
   "source": [
    "good_env = ['/data/homezvol1/freese/miniconda3/envs/SQANTI3.env',\\\n",
    "            '/share/crsp/lab/seyedam/freese/mortazavi_lab/data/230526_gencode',\\\n",
    "            '/data/homezvol1/freese/mortazavi_lab/bin/cDNA_Cupcake/sequence',\\\n",
    "            '/data/homezvol1/freese/mortazavi_lab/bin/cDNA_Cupcake',\\\n",
    "            '/data/homezvol1/freese/miniconda3/envs/SQANTI3.env/lib/python310.zip',\\\n",
    "            '/data/homezvol1/freese/miniconda3/envs/SQANTI3.env/lib/python3.10',\\\n",
    "            '/data/homezvol1/freese/miniconda3/envs/SQANTI3.env/lib/python3.10/lib-dynload',\\\n",
    "            '/data/homezvol1/freese/.local/lib/python3.10/site-packages',\\\n",
    "            '/dfs8/pub/freese/mortazavi_lab/bin/gencode_utr_fix',\\\n",
    "            '/data/homezvol1/freese/miniconda3/envs/SQANTI3.env/lib/python3.10/site-packages',\\\n",
    "            '/data/homezvol1/freese/miniconda3/envs/SQANTI3.env/lib/python3.10/site-packages/bcbio_gff-0.6.6-py2.7.egg',\\\n",
    "            '/data/homezvol1/freese/miniconda3/envs/SQANTI3.env/lib/python3.10/site-packages/cupcake-29.0.0-py3.10-linux-x86_64.egg',\\\n",
    "            '/data/homezvol1/freese/miniconda3/envs/SQANTI3.env/lib/python3.10/site-packages/sklearn-0.0.post5-py3.10.egg',\\\n",
    "            '/data/homezvol1/freese/miniconda3/envs/SQANTI3.env/lib/python3.10/site-packages/bcbio_gff-0.7.0-py3.10.egg']\n",
    "print(len(good_env))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7926390b-9834-40bd-b86b-edce8eb3e0b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11\n",
      "/data/homezvol1/freese/miniconda3/envs/SQANTI3.env/lib/python3.10/lib-dynload\n",
      "/data/homezvol1/freese/.local/lib/python3.10/site-packages\n",
      "/data/homezvol1/freese/miniconda3/envs/SQANTI3.env/lib/python3.10\n",
      "/data/homezvol1/freese/miniconda3/envs/SQANTI3.env\n",
      "/data/homezvol1/freese/miniconda3/envs/SQANTI3.env/lib/python310.zip\n",
      "/data/homezvol1/freese/miniconda3/envs/SQANTI3.env/lib/python3.10/site-packages\n",
      "/data/homezvol1/freese/miniconda3/envs/SQANTI3.env/lib/python3.10/site-packages/bcbio_gff-0.7.0-py3.10.egg\n",
      "/data/homezvol1/freese/miniconda3/envs/SQANTI3.env/lib/python3.10/site-packages/bcbio_gff-0.6.6-py2.7.egg\n",
      "/data/homezvol1/freese/miniconda3/envs/SQANTI3.env/lib/python3.10/site-packages/sklearn-0.0.post5-py3.10.egg\n",
      "/dfs8/pub/freese/mortazavi_lab/bin/gencode_utr_fix\n",
      "/data/homezvol1/freese/miniconda3/envs/SQANTI3.env/lib/python3.10/site-packages/cupcake-29.0.0-py3.10-linux-x86_64.egg\n"
     ]
    }
   ],
   "source": [
    "# commons\n",
    "common = list(set(good_env)&(set(bad_env)))\n",
    "print(len(common))\n",
    "for c in common:\n",
    "    print(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6190a674-21cd-45a8-8a48-3a61c2751dd9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n",
      "/data/homezvol1/freese/mortazavi_lab/bin/cDNA_Cupcake/sequence\n",
      "/share/crsp/lab/seyedam/freese/mortazavi_lab/data/230526_gencode\n",
      "/data/homezvol1/freese/mortazavi_lab/bin/cDNA_Cupcake\n"
     ]
    }
   ],
   "source": [
    "# only in good\n",
    "good = list(set(good_env)-(set(bad_env)))\n",
    "print(len(good))\n",
    "for c in good:\n",
    "    print(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63d0d1cd-e60b-46ca-a3e1-55cc104e8126",
   "metadata": {},
   "outputs": [],
   "source": [
    "# only in bad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c712f95f-c4da-41c3-aa99-6a886e926081",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "4e67ed0d-bbc3-4c01-ab06-9d7bc5f3eb65",
   "metadata": {},
   "source": [
    "## 230616 getting expression values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "50643803-bd7f-4a1d-8c00-4d737d041f34",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ab_from_gff(gff_file, ofile):\n",
    "    df = pr.read_gff(gff_file).as_df() \n",
    "    df = df[['transcript_id', 'flrpm', 'rpm']]\n",
    "    df = df.drop_duplicates()\n",
    "    assert len(df.loc[df.transcript_id.duplicated(keep=False)].index) == 0\n",
    "    df.to_csv(ofile, sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "67d5916a-8dc5-4db7-a054-d077ca3bce57",
   "metadata": {},
   "outputs": [],
   "source": [
    "gff_file = 'data/human/Brain_Embryo_ont_pre-capture_human_all.gff'\n",
    "ofile = 'data/human/Brain_Embryo_ont_pre-capture_human_all_abundance.tsv'\n",
    "get_ab_from_gff(gff_file, ofile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "71e03694-1f1f-4e96-948b-105c29bc047c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pr.read_gff('data/human/Brain_Embryo_ont_pre-capture_human_all.gff').as_df()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c638a395-e621-45a6-8e27-34040e96c0f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df[['transcript_id', 'flrpm', 'rpm']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "841f83be-b8b6-4cc6-b91a-da9c92dd7d4d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "213704\n",
      "51638\n"
     ]
    }
   ],
   "source": [
    "print(len(df.index))\n",
    "df = df.drop_duplicates()\n",
    "print(len(df.index))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "7a4e5d36-62b9-47ae-889e-ca8a403f6ea5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>flrpm</th>\n",
       "      <th>rpm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [transcript_id, flrpm, rpm]\n",
       "Index: []"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.transcript_id.duplicated(keep=False)].sort_values(by='transcript_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "8098a324-41cf-409b-af83-670679a384ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>flrpm</th>\n",
       "      <th>rpm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ont-Crg-CapTrap_HpreCap_0+_EmbBrain01Rep1.NAM_...</td>\n",
       "      <td>25.271982</td>\n",
       "      <td>25.271982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ont-Crg-CapTrap_HpreCap_0+_EmbBrain01Rep1.NAM_...</td>\n",
       "      <td>16.847988</td>\n",
       "      <td>33.695976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ont-Crg-CapTrap_HpreCap_0+_EmbBrain01Rep1.NAM_...</td>\n",
       "      <td>33.695976</td>\n",
       "      <td>623.375556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ont-Crg-CapTrap_HpreCap_0+_EmbBrain01Rep1.NAM_...</td>\n",
       "      <td>16.847988</td>\n",
       "      <td>210.59985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ont-Crg-CapTrap_HpreCap_0+_EmbBrain01Rep1.NAM_...</td>\n",
       "      <td>16.847988</td>\n",
       "      <td>75.815946</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        transcript_id      flrpm         rpm\n",
       "0   ont-Crg-CapTrap_HpreCap_0+_EmbBrain01Rep1.NAM_...  25.271982   25.271982\n",
       "3   ont-Crg-CapTrap_HpreCap_0+_EmbBrain01Rep1.NAM_...  16.847988   33.695976\n",
       "8   ont-Crg-CapTrap_HpreCap_0+_EmbBrain01Rep1.NAM_...  33.695976  623.375556\n",
       "9   ont-Crg-CapTrap_HpreCap_0+_EmbBrain01Rep1.NAM_...  16.847988   210.59985\n",
       "10  ont-Crg-CapTrap_HpreCap_0+_EmbBrain01Rep1.NAM_...  16.847988   75.815946"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c60722f5-172e-41e4-91d7-3e762a93f4e0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "2be3c7f3-04b3-4313-8df7-1dddca94f142",
   "metadata": {},
   "source": [
    "## 230615 input parsing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "cf266d86-4fa8-4b2a-8139-04ca371963a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_input_config(config, config_file, subset):\n",
    "    df = pd.read_csv(config_file, sep='\\t')\n",
    "    df['dataset'] = df.fname.str.split('.', n=1, expand=True)[0]\n",
    "    df['subset'] = np.nan\n",
    "    df.loc[df.fname.str.contains('all.gff.gz'), 'subset'] = 'all'\n",
    "    df.loc[df.fname.str.contains('cagePolyASupported.gff.gz'), 'subset'] = 'supported'\n",
    "\n",
    "    # add the metadata\n",
    "    species = ['human', 'mouse']\n",
    "    meta_df = pd.DataFrame()\n",
    "    for s in species:\n",
    "        temp = pd.read_csv(expand(config['data']['meta'], species=s)[0],\n",
    "                           sep='\\t')\n",
    "        temp['species'] = s\n",
    "        meta_df = pd.concat([meta_df, temp], axis=0)\n",
    "\n",
    "    df = df.merge(meta_df, on='dataset', how='left')\n",
    "    df['temp'] = df['tissue']+'_'+\\\n",
    "                 df['age']+'_'+\\\n",
    "                 df['platform']+'_'+\\\n",
    "                 df['capture']+'_'+\\\n",
    "                 df['species']+'_'+\\\n",
    "                 df['subset']\n",
    "    \n",
    "    df = df.loc[df.subset==subset]\n",
    "    print(len(df.temp.unique()))\n",
    "    # print(df.head())\n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "b0dc4d58-1bda-47cb-8421-2fabdda2b23c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "88\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fname</th>\n",
       "      <th>path</th>\n",
       "      <th>dataset</th>\n",
       "      <th>subset</th>\n",
       "      <th>tissue</th>\n",
       "      <th>age</th>\n",
       "      <th>platform</th>\n",
       "      <th>capture</th>\n",
       "      <th>rep1</th>\n",
       "      <th>rep2</th>\n",
       "      <th>species</th>\n",
       "      <th>temp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIDMWbEPP0101.splicing_status-all.endSupport-a...</td>\n",
       "      <td>https://public-docs.crg.es/rguigo/Data/gkaur/L...</td>\n",
       "      <td>SIDMWbEPP0101</td>\n",
       "      <td>all</td>\n",
       "      <td>ESC</td>\n",
       "      <td>Embryo</td>\n",
       "      <td>pacBioSII</td>\n",
       "      <td>pre-capture</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>mouse</td>\n",
       "      <td>ESC_Embryo_pacBioSII_pre-capture_mouse_all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIDMWbEPC0101.splicing_status-all.endSupport-a...</td>\n",
       "      <td>https://public-docs.crg.es/rguigo/Data/gkaur/L...</td>\n",
       "      <td>SIDMWbEPC0101</td>\n",
       "      <td>all</td>\n",
       "      <td>ESC</td>\n",
       "      <td>Embryo</td>\n",
       "      <td>pacBioSII</td>\n",
       "      <td>post-capture</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>mouse</td>\n",
       "      <td>ESC_Embryo_pacBioSII_post-capture_mouse_all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SIDMWbEOP0101.splicing_status-all.endSupport-a...</td>\n",
       "      <td>https://public-docs.crg.es/rguigo/Data/gkaur/L...</td>\n",
       "      <td>SIDMWbEOP0101</td>\n",
       "      <td>all</td>\n",
       "      <td>ESC</td>\n",
       "      <td>Embryo</td>\n",
       "      <td>ont</td>\n",
       "      <td>pre-capture</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>mouse</td>\n",
       "      <td>ESC_Embryo_ont_pre-capture_mouse_all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SIDMWbEOC0101.splicing_status-all.endSupport-a...</td>\n",
       "      <td>https://public-docs.crg.es/rguigo/Data/gkaur/L...</td>\n",
       "      <td>SIDMWbEOC0101</td>\n",
       "      <td>all</td>\n",
       "      <td>ESC</td>\n",
       "      <td>Embryo</td>\n",
       "      <td>ont</td>\n",
       "      <td>post-capture</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>mouse</td>\n",
       "      <td>ESC_Embryo_ont_post-capture_mouse_all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SIDMWbAPP0101.splicing_status-all.endSupport-a...</td>\n",
       "      <td>https://public-docs.crg.es/rguigo/Data/gkaur/L...</td>\n",
       "      <td>SIDMWbAPP0101</td>\n",
       "      <td>all</td>\n",
       "      <td>WBlood</td>\n",
       "      <td>Adult</td>\n",
       "      <td>pacBioSII</td>\n",
       "      <td>pre-capture</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>mouse</td>\n",
       "      <td>WBlood_Adult_pacBioSII_pre-capture_mouse_all</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               fname  \\\n",
       "1  SIDMWbEPP0101.splicing_status-all.endSupport-a...   \n",
       "3  SIDMWbEPC0101.splicing_status-all.endSupport-a...   \n",
       "5  SIDMWbEOP0101.splicing_status-all.endSupport-a...   \n",
       "7  SIDMWbEOC0101.splicing_status-all.endSupport-a...   \n",
       "9  SIDMWbAPP0101.splicing_status-all.endSupport-a...   \n",
       "\n",
       "                                                path        dataset subset  \\\n",
       "1  https://public-docs.crg.es/rguigo/Data/gkaur/L...  SIDMWbEPP0101    all   \n",
       "3  https://public-docs.crg.es/rguigo/Data/gkaur/L...  SIDMWbEPC0101    all   \n",
       "5  https://public-docs.crg.es/rguigo/Data/gkaur/L...  SIDMWbEOP0101    all   \n",
       "7  https://public-docs.crg.es/rguigo/Data/gkaur/L...  SIDMWbEOC0101    all   \n",
       "9  https://public-docs.crg.es/rguigo/Data/gkaur/L...  SIDMWbAPP0101    all   \n",
       "\n",
       "   tissue     age   platform       capture  rep1  rep2 species  \\\n",
       "1     ESC  Embryo  pacBioSII   pre-capture     1     1   mouse   \n",
       "3     ESC  Embryo  pacBioSII  post-capture     1     1   mouse   \n",
       "5     ESC  Embryo        ont   pre-capture     1     1   mouse   \n",
       "7     ESC  Embryo        ont  post-capture     1     1   mouse   \n",
       "9  WBlood   Adult  pacBioSII   pre-capture     1     1   mouse   \n",
       "\n",
       "                                           temp  \n",
       "1    ESC_Embryo_pacBioSII_pre-capture_mouse_all  \n",
       "3   ESC_Embryo_pacBioSII_post-capture_mouse_all  \n",
       "5          ESC_Embryo_ont_pre-capture_mouse_all  \n",
       "7         ESC_Embryo_ont_post-capture_mouse_all  \n",
       "9  WBlood_Adult_pacBioSII_pre-capture_mouse_all  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = parse_input_config(config, '230614_config.tsv', 'all')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "48873a40-25b0-4470-bed2-ea6e92672ebf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "176\n",
      "88\n",
      "176\n",
      "                                               fname  \\\n",
      "0  SIDMWbEPP0101.splicing_status-all.endSupport-c...   \n",
      "1  SIDMWbEPP0101.splicing_status-all.endSupport-a...   \n",
      "2  SIDMWbEPC0101.splicing_status-all.endSupport-c...   \n",
      "3  SIDMWbEPC0101.splicing_status-all.endSupport-a...   \n",
      "4  SIDMWbEOP0101.splicing_status-all.endSupport-c...   \n",
      "\n",
      "                                                path        dataset  \\\n",
      "0  https://public-docs.crg.es/rguigo/Data/gkaur/L...  SIDMWbEPP0101   \n",
      "1  https://public-docs.crg.es/rguigo/Data/gkaur/L...  SIDMWbEPP0101   \n",
      "2  https://public-docs.crg.es/rguigo/Data/gkaur/L...  SIDMWbEPC0101   \n",
      "3  https://public-docs.crg.es/rguigo/Data/gkaur/L...  SIDMWbEPC0101   \n",
      "4  https://public-docs.crg.es/rguigo/Data/gkaur/L...  SIDMWbEOP0101   \n",
      "\n",
      "      subset tissue     age   platform       capture  rep1  rep2 species  \\\n",
      "0  supported    ESC  Embryo  pacBioSII   pre-capture     1     1   mouse   \n",
      "1        all    ESC  Embryo  pacBioSII   pre-capture     1     1   mouse   \n",
      "2  supported    ESC  Embryo  pacBioSII  post-capture     1     1   mouse   \n",
      "3        all    ESC  Embryo  pacBioSII  post-capture     1     1   mouse   \n",
      "4  supported    ESC  Embryo        ont   pre-capture     1     1   mouse   \n",
      "\n",
      "                                                temp  \n",
      "0   ESC_Embryo_pacBioSII_pre-capture_mouse_supported  \n",
      "1         ESC_Embryo_pacBioSII_pre-capture_mouse_all  \n",
      "2  ESC_Embryo_pacBioSII_post-capture_mouse_supported  \n",
      "3        ESC_Embryo_pacBioSII_post-capture_mouse_all  \n",
      "4         ESC_Embryo_ont_pre-capture_mouse_supported  \n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv('230614_config.tsv', sep='\\t')\n",
    "df['dataset_id'] = df.fname.str.split('.', n=1, expand=True)[0]\n",
    "df['subset'] = np.nan\n",
    "df.loc[df.fname.str.contains('all.gff.gz'), 'subset'] = 'all'\n",
    "df.loc[df.fname.str.contains('cagePolyASupported.gff.gz'), 'subset'] = 'supported'\n",
    "\n",
    "# add the metadata\n",
    "species = ['human', 'mouse']\n",
    "meta_df = pd.DataFrame()\n",
    "for s in species:\n",
    "    temp = pd.read_csv(expand(config['data']['meta'], species=s)[0],\n",
    "                       sep='\\t')\n",
    "    temp['species'] = s\n",
    "    meta_df = pd.concat([meta_df, temp], axis=0)\n",
    "\n",
    "df = df.merge(meta_df, on='dataset', how='left')\n",
    "print(len(df.index))\n",
    "print(len(df.dataset.unique()))\n",
    "df['dataset'] = df['tissue']+'_'+\\\n",
    "             df['age']+'_'+\\\n",
    "             df['platform']+'_'+\\\n",
    "             df['capture']+'_'+\\\n",
    "             df['species']+'_'+\\\n",
    "             df['subset']\n",
    "print(len(df.temp.unique()))\n",
    "print(df.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "4d831835-1743-4ded-bb04-a788fc56cc89",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['link'] = df.path+df.fname"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "1f449b62-2a4a-496c-a51c-9236c4c44f5d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fname</th>\n",
       "      <th>path</th>\n",
       "      <th>dataset</th>\n",
       "      <th>subset</th>\n",
       "      <th>tissue</th>\n",
       "      <th>age</th>\n",
       "      <th>platform</th>\n",
       "      <th>capture</th>\n",
       "      <th>rep1</th>\n",
       "      <th>rep2</th>\n",
       "      <th>species</th>\n",
       "      <th>temp</th>\n",
       "      <th>link</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SIDMWbEPP0101.splicing_status-all.endSupport-c...</td>\n",
       "      <td>https://public-docs.crg.es/rguigo/Data/gkaur/L...</td>\n",
       "      <td>SIDMWbEPP0101</td>\n",
       "      <td>supported</td>\n",
       "      <td>ESC</td>\n",
       "      <td>Embryo</td>\n",
       "      <td>pacBioSII</td>\n",
       "      <td>pre-capture</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>mouse</td>\n",
       "      <td>ESC_Embryo_pacBioSII_pre-capture_mouse</td>\n",
       "      <td>https://public-docs.crg.es/rguigo/Data/gkaur/L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SIDMWbEPP0101.splicing_status-all.endSupport-a...</td>\n",
       "      <td>https://public-docs.crg.es/rguigo/Data/gkaur/L...</td>\n",
       "      <td>SIDMWbEPP0101</td>\n",
       "      <td>all</td>\n",
       "      <td>ESC</td>\n",
       "      <td>Embryo</td>\n",
       "      <td>pacBioSII</td>\n",
       "      <td>pre-capture</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>mouse</td>\n",
       "      <td>ESC_Embryo_pacBioSII_pre-capture_mouse</td>\n",
       "      <td>https://public-docs.crg.es/rguigo/Data/gkaur/L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SIDMWbEPC0101.splicing_status-all.endSupport-c...</td>\n",
       "      <td>https://public-docs.crg.es/rguigo/Data/gkaur/L...</td>\n",
       "      <td>SIDMWbEPC0101</td>\n",
       "      <td>supported</td>\n",
       "      <td>ESC</td>\n",
       "      <td>Embryo</td>\n",
       "      <td>pacBioSII</td>\n",
       "      <td>post-capture</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>mouse</td>\n",
       "      <td>ESC_Embryo_pacBioSII_post-capture_mouse</td>\n",
       "      <td>https://public-docs.crg.es/rguigo/Data/gkaur/L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SIDMWbEPC0101.splicing_status-all.endSupport-a...</td>\n",
       "      <td>https://public-docs.crg.es/rguigo/Data/gkaur/L...</td>\n",
       "      <td>SIDMWbEPC0101</td>\n",
       "      <td>all</td>\n",
       "      <td>ESC</td>\n",
       "      <td>Embryo</td>\n",
       "      <td>pacBioSII</td>\n",
       "      <td>post-capture</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>mouse</td>\n",
       "      <td>ESC_Embryo_pacBioSII_post-capture_mouse</td>\n",
       "      <td>https://public-docs.crg.es/rguigo/Data/gkaur/L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SIDMWbEOP0101.splicing_status-all.endSupport-c...</td>\n",
       "      <td>https://public-docs.crg.es/rguigo/Data/gkaur/L...</td>\n",
       "      <td>SIDMWbEOP0101</td>\n",
       "      <td>supported</td>\n",
       "      <td>ESC</td>\n",
       "      <td>Embryo</td>\n",
       "      <td>ont</td>\n",
       "      <td>pre-capture</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>mouse</td>\n",
       "      <td>ESC_Embryo_ont_pre-capture_mouse</td>\n",
       "      <td>https://public-docs.crg.es/rguigo/Data/gkaur/L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>SIDHBrAPC0301.splicing_status-all.endSupport-a...</td>\n",
       "      <td>https://public-docs.crg.es/rguigo/Data/gkaur/L...</td>\n",
       "      <td>SIDHBrAPC0301</td>\n",
       "      <td>all</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Adult</td>\n",
       "      <td>pacBioSII</td>\n",
       "      <td>post-capture</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>human</td>\n",
       "      <td>Brain_Adult_pacBioSII_post-capture_human</td>\n",
       "      <td>https://public-docs.crg.es/rguigo/Data/gkaur/L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>SIDHBrAOP0301.splicing_status-all.endSupport-c...</td>\n",
       "      <td>https://public-docs.crg.es/rguigo/Data/gkaur/L...</td>\n",
       "      <td>SIDHBrAOP0301</td>\n",
       "      <td>supported</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Adult</td>\n",
       "      <td>ont</td>\n",
       "      <td>pre-capture</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>human</td>\n",
       "      <td>Brain_Adult_ont_pre-capture_human</td>\n",
       "      <td>https://public-docs.crg.es/rguigo/Data/gkaur/L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>SIDHBrAOP0301.splicing_status-all.endSupport-a...</td>\n",
       "      <td>https://public-docs.crg.es/rguigo/Data/gkaur/L...</td>\n",
       "      <td>SIDHBrAOP0301</td>\n",
       "      <td>all</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Adult</td>\n",
       "      <td>ont</td>\n",
       "      <td>pre-capture</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>human</td>\n",
       "      <td>Brain_Adult_ont_pre-capture_human</td>\n",
       "      <td>https://public-docs.crg.es/rguigo/Data/gkaur/L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>SIDHBrAOC0301.splicing_status-all.endSupport-c...</td>\n",
       "      <td>https://public-docs.crg.es/rguigo/Data/gkaur/L...</td>\n",
       "      <td>SIDHBrAOC0301</td>\n",
       "      <td>supported</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Adult</td>\n",
       "      <td>ont</td>\n",
       "      <td>post-capture</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>human</td>\n",
       "      <td>Brain_Adult_ont_post-capture_human</td>\n",
       "      <td>https://public-docs.crg.es/rguigo/Data/gkaur/L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>SIDHBrAOC0301.splicing_status-all.endSupport-a...</td>\n",
       "      <td>https://public-docs.crg.es/rguigo/Data/gkaur/L...</td>\n",
       "      <td>SIDHBrAOC0301</td>\n",
       "      <td>all</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Adult</td>\n",
       "      <td>ont</td>\n",
       "      <td>post-capture</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>human</td>\n",
       "      <td>Brain_Adult_ont_post-capture_human</td>\n",
       "      <td>https://public-docs.crg.es/rguigo/Data/gkaur/L...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>176 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 fname  \\\n",
       "0    SIDMWbEPP0101.splicing_status-all.endSupport-c...   \n",
       "1    SIDMWbEPP0101.splicing_status-all.endSupport-a...   \n",
       "2    SIDMWbEPC0101.splicing_status-all.endSupport-c...   \n",
       "3    SIDMWbEPC0101.splicing_status-all.endSupport-a...   \n",
       "4    SIDMWbEOP0101.splicing_status-all.endSupport-c...   \n",
       "..                                                 ...   \n",
       "171  SIDHBrAPC0301.splicing_status-all.endSupport-a...   \n",
       "172  SIDHBrAOP0301.splicing_status-all.endSupport-c...   \n",
       "173  SIDHBrAOP0301.splicing_status-all.endSupport-a...   \n",
       "174  SIDHBrAOC0301.splicing_status-all.endSupport-c...   \n",
       "175  SIDHBrAOC0301.splicing_status-all.endSupport-a...   \n",
       "\n",
       "                                                  path        dataset  \\\n",
       "0    https://public-docs.crg.es/rguigo/Data/gkaur/L...  SIDMWbEPP0101   \n",
       "1    https://public-docs.crg.es/rguigo/Data/gkaur/L...  SIDMWbEPP0101   \n",
       "2    https://public-docs.crg.es/rguigo/Data/gkaur/L...  SIDMWbEPC0101   \n",
       "3    https://public-docs.crg.es/rguigo/Data/gkaur/L...  SIDMWbEPC0101   \n",
       "4    https://public-docs.crg.es/rguigo/Data/gkaur/L...  SIDMWbEOP0101   \n",
       "..                                                 ...            ...   \n",
       "171  https://public-docs.crg.es/rguigo/Data/gkaur/L...  SIDHBrAPC0301   \n",
       "172  https://public-docs.crg.es/rguigo/Data/gkaur/L...  SIDHBrAOP0301   \n",
       "173  https://public-docs.crg.es/rguigo/Data/gkaur/L...  SIDHBrAOP0301   \n",
       "174  https://public-docs.crg.es/rguigo/Data/gkaur/L...  SIDHBrAOC0301   \n",
       "175  https://public-docs.crg.es/rguigo/Data/gkaur/L...  SIDHBrAOC0301   \n",
       "\n",
       "        subset tissue     age   platform       capture  rep1  rep2 species  \\\n",
       "0    supported    ESC  Embryo  pacBioSII   pre-capture     1     1   mouse   \n",
       "1          all    ESC  Embryo  pacBioSII   pre-capture     1     1   mouse   \n",
       "2    supported    ESC  Embryo  pacBioSII  post-capture     1     1   mouse   \n",
       "3          all    ESC  Embryo  pacBioSII  post-capture     1     1   mouse   \n",
       "4    supported    ESC  Embryo        ont   pre-capture     1     1   mouse   \n",
       "..         ...    ...     ...        ...           ...   ...   ...     ...   \n",
       "171        all  Brain   Adult  pacBioSII  post-capture     3     1   human   \n",
       "172  supported  Brain   Adult        ont   pre-capture     3     1   human   \n",
       "173        all  Brain   Adult        ont   pre-capture     3     1   human   \n",
       "174  supported  Brain   Adult        ont  post-capture     3     1   human   \n",
       "175        all  Brain   Adult        ont  post-capture     3     1   human   \n",
       "\n",
       "                                         temp  \\\n",
       "0      ESC_Embryo_pacBioSII_pre-capture_mouse   \n",
       "1      ESC_Embryo_pacBioSII_pre-capture_mouse   \n",
       "2     ESC_Embryo_pacBioSII_post-capture_mouse   \n",
       "3     ESC_Embryo_pacBioSII_post-capture_mouse   \n",
       "4            ESC_Embryo_ont_pre-capture_mouse   \n",
       "..                                        ...   \n",
       "171  Brain_Adult_pacBioSII_post-capture_human   \n",
       "172         Brain_Adult_ont_pre-capture_human   \n",
       "173         Brain_Adult_ont_pre-capture_human   \n",
       "174        Brain_Adult_ont_post-capture_human   \n",
       "175        Brain_Adult_ont_post-capture_human   \n",
       "\n",
       "                                                  link  \n",
       "0    https://public-docs.crg.es/rguigo/Data/gkaur/L...  \n",
       "1    https://public-docs.crg.es/rguigo/Data/gkaur/L...  \n",
       "2    https://public-docs.crg.es/rguigo/Data/gkaur/L...  \n",
       "3    https://public-docs.crg.es/rguigo/Data/gkaur/L...  \n",
       "4    https://public-docs.crg.es/rguigo/Data/gkaur/L...  \n",
       "..                                                 ...  \n",
       "171  https://public-docs.crg.es/rguigo/Data/gkaur/L...  \n",
       "172  https://public-docs.crg.es/rguigo/Data/gkaur/L...  \n",
       "173  https://public-docs.crg.es/rguigo/Data/gkaur/L...  \n",
       "174  https://public-docs.crg.es/rguigo/Data/gkaur/L...  \n",
       "175  https://public-docs.crg.es/rguigo/Data/gkaur/L...  \n",
       "\n",
       "[176 rows x 13 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datasets = df.dataset.tolist()\n",
    "species = df.species.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "341f1a76-55dc-4a48-8765-c9f63620f138",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95473326-dad9-429c-8684-ab8c2bd1db03",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c3013c2e-d9f7-4a34-8d45-650a4bbf4fcc",
   "metadata": {},
   "outputs": [],
   "source": [
    "config_file = 'config.yml'\n",
    "with open(config_file) as f:\n",
    "    config = yaml.safe_load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8c598b42-ab30-4f80-a37b-f9850e1bd595",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'data/{species}/bepis_{end_type}_agg.bed'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "config['data']['cerberus']['agg_ends']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5c4a0cde-4bdb-4fcb-a334-fbf569fb7421",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['data/human/bepis_tes_agg.bed']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expand(                         config['data']['cerberus']['agg_ends'],\n",
    "       zip,\n",
    "                                species='human',\n",
    "                                end_type='tes',\n",
    "                                allow_missing=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0d64031-4673-4f1d-ab3e-bc65bb86630e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36f4ccd6-174a-47c2-a2b0-44ab34d7f38e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
